buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java

```diff
@@ -0,0 +1,187 @@
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle;
+
+import groovy.lang.Closure;
+import org.gradle.api.Action;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.Task;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ConfigurationContainer;
+import org.gradle.api.artifacts.dsl.DependencyHandler;
+import org.gradle.api.file.CopySpec;
+import org.gradle.api.file.FileCopyDetails;
+import org.gradle.api.file.FileTree;
+import org.gradle.api.file.RelativePath;
+import org.gradle.api.plugins.ExtraPropertiesExtension;
+import org.gradle.api.tasks.Copy;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.function.Supplier;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class JdkDownloadPlugin implements Plugin<Project> {
+
+    private static final Pattern JDK_VERSION = Pattern.compile("(\\d+)(\\.\\d+\\.\\d+)?\\+(\\d+)");
+
+    private List<String> ALLOWED_PLATFORMS = Collections.unmodifiableList(Arrays.asList("linux", "windows", "darwin"));
+
+    @Override
+    public void apply(Project project) {
+
+        // create usesJdk task "method"
+        project.getTasks().all(task ->
+            task.getExtensions().findByType(ExtraPropertiesExtension.class).set("usesJdk", new Closure<Void>(project, task) {
+                public void doCall(String version, String platform) {
+                    if (version == null) {
+                        throw new IllegalArgumentException("version must be specified to usesJdk");
+                    }
+                    if (platform == null) {
+                        throw new IllegalArgumentException("platform must be specified to usesJdk");
+                    }
+                    if (ALLOWED_PLATFORMS.contains(platform) == false) {
+                        throw new IllegalArgumentException("platform must be one of " + ALLOWED_PLATFORMS);
+                    }
+                    ExtraPropertiesExtension extraProperties = task.getExtensions().findByType(ExtraPropertiesExtension.class);
+                    if (extraProperties.has("jdkHome")) {
+                        throw new IllegalArgumentException("jdk version already set for task");
```

Comment:
Maybe include in the exception what version was set previously

Reply:
This is actually difficult to figure out, since the version is not directly stored anywhere, it is only part of various configuration names etc.

Reply:
maybe `extraProperties.get("jdkHome")` would help ?

Reply:
It is just an arbitrary object, and can't be resolved to toString until runtime, while this is at configuration time.

Reply:
ok, let's leave it for now, we can revisit if it becomes too painful to debug